src/Http/Routing/src/Builder/MapActionEndpointRouteBuilderExtensions.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Internal;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    /// <summary>
+    /// Provides extension methods for <see cref="IEndpointRouteBuilder"/> to define HTTP API endpoints.
+    /// </summary>
+    public static class MapActionEndpointRouteBuilderExtensions
+    {
+        /// <summary>
+        /// Adds a <see cref="RouteEndpoint"/> to the <see cref="IEndpointRouteBuilder"/> that matches the pattern specified via attributes.
+        /// </summary>
+        /// <param name="endpoints">The <see cref="IEndpointRouteBuilder"/> to add the route to.</param>
+        /// <param name="action">The delegate executed when the endpoint is matched.</param>
+        /// <returns>An <see cref="IEndpointConventionBuilder"/> that can be used to further customize the endpoint.</returns>
+        public static MapActionEndpointConventionBuilder MapAction(
+            this IEndpointRouteBuilder endpoints,
+            Delegate action)
+        {
+            if (endpoints is null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            if (action is null)
+            {
+                throw new ArgumentNullException(nameof(action));
+            }
+
+            var requestDelegate = MapActionExpressionTreeBuilder.BuildRequestDelegate(action);
+
+            var routeAttributes = action.Method.GetCustomAttributes().OfType<IRoutePatternMetadata>();
+            var conventionBuilders = new List<IEndpointConventionBuilder>();
+
+            const int defaultOrder = 0;
+
+            foreach (var routeAttribute in routeAttributes)
```

Comment:
This is super-de-duper simplified compared to what MVC's attributes **mean**. You as mentioned elsewhere, it's probs a good idea to limit what's possible to express with the attribute to some reasonable set. You could throw here, or have an analyzer.

Reply:
I'm also not saying it should be as complex as what MVC allows - it should be simpler than that 😆

Reply:
Would it be problematic to say 

a) one action -> one route attribute (exactly one method constraint \ template)
b) each route attribute must have a template

The current implementation simply skips over route attributes without a template.